peer_client.go

```diff
@@ -470,31 +402,26 @@ func (c *PeerClient) sendBatch(ctx context.Context, queue []*request) {
 	}
 }
 
-// Shutdown will gracefully shutdown the client connection, until the context is cancelled
+// Shutdown waits until all outstanding requests have finished or the context is cancelled.
+// Then it closes the grpc connection.
 func (c *PeerClient) Shutdown(ctx context.Context) error {
-	// Take the write lock since we're going to modify the closing state
-	c.mutex.Lock()
-	if c.status == peerClosing || c.status == peerNotConnected {
-		c.mutex.Unlock()
-		return nil
-	}
-	defer c.mutex.Unlock()
-
-	c.status = peerClosing
+	// ensure we don't leak goroutines, even if the Shutdown times out
+	defer c.conn.Close()
 
-	defer func() {
-		if c.conn != nil {
-			c.conn.Close()
-		}
-	}()
-
-	// This allows us to wait on the waitgroup, or until the context
-	// has been cancelled. This doesn't leak goroutines, because
-	// closing the connection will kill any outstanding requests.
 	waitChan := make(chan struct{})
 	go func() {
+		// drain in-flight requests
+		c.wgMutex.Lock()
+		defer c.wgMutex.Unlock()
 		c.wg.Wait()
+
+		// clear errors
+		c.lastErrs = collections.NewLRUCache(100)
```

Comment:
not sure if we still want this 🤔

Reply:
We can punt on this.  It clears the client's error list and there could be a good reason to do this.